File: CS/PersonalOrganizer/Common/DataModel/EntityFramework/DbUnitOfWork.cs
using System;
using System.Collections.Generic;
using System.Data.Entity;
using System.Data.Entity.Validation;
using System.Data.Entity.Infrastructure;
using System.Linq;

namespace PersonalOrganizer.Common.DataModel.EntityFramework {
    public abstract class DbUnitOfWork<TContext> : IUnitOfWork where TContext : DbContext {
        public DbUnitOfWork(TContext context) {
            Context = context;
        }
        public TContext Context { get; private set; }
        void IUnitOfWork.SaveChanges() {
            try {
                Context.SaveChanges();
            } catch(DbEntityValidationException ex) {
                throw DbExceptionsConverter.Convert(ex);
            } catch(DbUpdateException ex) {
                throw DbExceptionsConverter.Convert(ex);
            }
        }
        EntityState IUnitOfWork.GetState(object entity) {
            return GetEntityState(Context.Entry(entity).State);
        }
        void IUnitOfWork.Update(object entity) { }
        void IUnitOfWork.Detach(object entity) {
            Context.Entry(entity).State = System.Data.Entity.EntityState.Detached;
        }
        EntityState GetEntityState(System.Data.Entity.EntityState entityStates) {
            switch(entityStates) {
                case System.Data.Entity.EntityState.Added:
                    return EntityState.Added;
                case System.Data.Entity.EntityState.Deleted:
                    return EntityState.Deleted;
                case System.Data.Entity.EntityState.Detached:
                    return EntityState.Detached;
                case System.Data.Entity.EntityState.Modified:
                    return EntityState.Modified;
                case System.Data.Entity.EntityState.Unchanged:
                    return EntityState.Unchanged;
                default:
                    throw new NotImplementedException();
            }
        }
    }
}

File: VB/PersonalOrganizer/Common/DataModel/EntityFramework/DbUnitOfWork.vb
Imports System
Imports System.Collections.Generic
Imports System.Data.Entity
Imports System.Data.Entity.Validation
Imports System.Data.Entity.Infrastructure
Imports System.Linq
Namespace Common.DataModel.EntityFramework
    Public MustInherit Class DbUnitOfWork(Of TContext As DbContext)
        Implements IUnitOfWork
        Private _Context As TContext
        Public Sub New(ByVal context As TContext)
            Me._Context = context
        End Sub
        Public ReadOnly Property Context As TContext
            Get
                Return _Context
            End Get
        End Property
        Private Sub SaveChanges() Implements IUnitOfWork.SaveChanges
            Try
                Context.SaveChanges()
            Catch ex As DbEntityValidationException
                Throw DbExceptionsConverter.Convert(ex)
            Catch ex As DbUpdateException
                Throw DbExceptionsConverter.Convert(ex)
            End Try
        End Sub
        Private Function GetState(ByVal entity As Object) As EntityState Implements IUnitOfWork.GetState
            Return GetEntityState(Context.Entry(entity).State)
        End Function
        Private Sub Update(ByVal entity As Object) Implements IUnitOfWork.Update
        End Sub
        Private Sub Detach(ByVal entity As Object) Implements IUnitOfWork.Detach
            Context.Entry(entity).State = System.Data.Entity.EntityState.Detached
        End Sub
        Private Function GetEntityState(ByVal entityStates As System.Data.Entity.EntityState) As EntityState
            Select Case entityStates
                Case System.Data.Entity.EntityState.Added
                    Return EntityState.Added
                Case System.Data.Entity.EntityState.Deleted
                    Return EntityState.Deleted
                Case System.Data.Entity.EntityState.Detached
                    Return EntityState.Detached
                Case System.Data.Entity.EntityState.Modified
                    Return EntityState.Modified
                Case System.Data.Entity.EntityState.Unchanged
                    Return EntityState.Unchanged
                Case Else
                    Throw New NotImplementedException()
            End Select
        End Function
    End Class
End Namespace
